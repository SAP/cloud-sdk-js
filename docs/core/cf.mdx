---
id: sap-cloud-abstractions
title: Feature title
hide_title: false
hide_table_of_contents: false
sidebar_label: Shiny label :)
description: Who should I read this?!
keywords:
- docs
- docusaurus
---
