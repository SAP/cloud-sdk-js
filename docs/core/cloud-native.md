---
id: cloud-native
title: Cloud native development concept
hide_title: false
hide_table_of_contents: false
sidebar_label: Cloud native
description: Why should I read this?!
keywords:
- docs
- docusaurus
image: https://myimagesrv.com/img.png
---
