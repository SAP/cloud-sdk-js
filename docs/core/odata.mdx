---
id: odata-api
title: OData API main concepts
hide_title: false
hide_table_of_contents: false
sidebar_label: OData API
description: OData API is a backbone for extending most of SAP provide services.
keywords:
- odata
- sap
- api
- cloud
- sdk
---
