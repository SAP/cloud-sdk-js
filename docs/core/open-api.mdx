---
id: open-api
title: Open API withing SAP infastructure
hide_title: false
hide_table_of_contents: false
sidebar_label: Open API
description: Leverage flexibility of Open API specificaion with SAP services by using our of the box Cloud SDK generator
keywords:
- docs
- docusaurus
image: https://myimagesrv.com/img.png
---
