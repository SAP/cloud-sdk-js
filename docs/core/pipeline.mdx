---
id: continuous-delivery
title: CI/CD and DevOps
hide_title: false
hide_table_of_contents: false
sidebar_label: Continuous Delivery (CI/CD)
description: Who should I read this?!
keywords:
- docs
- docusaurus
image: https://myimagesrv.com/img.png
---
