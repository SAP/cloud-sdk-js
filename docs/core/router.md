---
id: app-router
title: Cloud application router
hide_title: false
hide_table_of_contents: false
sidebar_label: Cloud router
description: Router is ubiquitos part of every Cloud native application.
keywords:
- router
- cloud
- sap
image: https://myimagesrv.com/img.png
---
