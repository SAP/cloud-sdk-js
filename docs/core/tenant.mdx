---
id: tenant
title: Multitenancy explained in simple words
hide_title: false
hide_table_of_contents: false
sidebar_label: Shiny label :)
description: When you App has to serve multiple consumers and fully separate their context multitenancy comes into play.
keywords:
- multitenancy
- tenant
- cloud
- sdk
image: https://myimagesrv.com/img.png
---
