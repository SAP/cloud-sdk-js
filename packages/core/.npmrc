@sap:registry=https://npm.sap.com
