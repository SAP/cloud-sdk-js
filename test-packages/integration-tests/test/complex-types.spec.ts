/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import { asc, Destination } from '@sap-cloud-sdk/core';
import {
  TestComplexType,
  TestEntity
} from '@sap-cloud-sdk/test-services/test-service';
import BigNumber from 'bignumber.js';
import nock from 'nock';
import { testEntityCollectionResponse } from './test-data/test-entity-collection-response';
import { basicCredentials } from './test-util/destination-encoder';

const servicePath = '/sap/opu/odata/sap/API_TEST_SRV';
const entityName = TestEntity._entityName;

const getAllResponse = testEntityCollectionResponse();

const destination: Destination = {
  url: 'https://example.com',
  username: 'username',
  password: 'password',
  sapClient: '123',
  authTokens: [],
  originalProperties: {}
};

describe('Complex types', () => {
  it('should allow filtering on complex type properties', async () => {
    nock(destination.url, {
      reqheaders: {
        authorization: basicCredentials(destination),
        accept: 'application/json',
        'content-type': 'application/json'
      }
    })
      .get(
        `${servicePath}/${entityName}?$format=json&$filter=(ComplexTypeProperty/StringProperty%20eq%20%27someComplexTypeProperty%27)`
      )
      .reply(200, getAllResponse);

    const request = TestEntity.requestBuilder()
      .getAll()
      .filter(
        TestEntity.COMPLEX_TYPE_PROPERTY.stringProperty.equals(
          'someComplexTypeProperty'
        )
      )
      .execute(destination);

    await expect(request).resolves.not.toThrow();
  });

  it('should allow ordering on complex type properties', async () => {
    nock(destination.url, {
      reqheaders: {
        authorization: basicCredentials(destination),
        accept: 'application/json',
        'content-type': 'application/json'
      }
    })
      .get(
        `${servicePath}/${entityName}?$format=json&$orderby=ComplexTypeProperty/StringProperty%20asc`
      )
      .reply(200, getAllResponse);

    const request = TestEntity.requestBuilder()
      .getAll()
      .orderBy(asc(TestEntity.COMPLEX_TYPE_PROPERTY.stringProperty))
      .execute(destination);

    await expect(request).resolves.not.toThrow();
  });

  it('should be constructable by a builder', () => {
    const actual = TestComplexType.build({
      StringProperty: 'random value',
      BooleanProperty: false,
      GuidProperty: 'aaaabbbb-aaaa-bbbb-aaaa-bbbbaaaabbbb',
      Int16Property: 4,
      Int32Property: 6,
      Int64Property: '54',
      TimeProperty: 'PT11H43M43S'
    });

    const expected = {
      stringProperty: 'random value',
      booleanProperty: false,
      guidProperty: 'aaaabbbb-aaaa-bbbb-aaaa-bbbbaaaabbbb',
      int16Property: 4,
      int32Property: 6,
      int64Property: new BigNumber(54),
      timeProperty: { hours: 11, minutes: 43, seconds: 43 }
    };

    expect(actual).toMatchObject(expected);
  });

  it('should fail when passing unknown properties to the builder', () => {
    expect(() =>
      TestComplexType.build({ SomethingElse: 'Fails!' })
    ).toThrowErrorMatchingSnapshot();
  });
});
