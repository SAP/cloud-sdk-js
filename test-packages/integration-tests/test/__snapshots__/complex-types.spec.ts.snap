// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complex types should fail when passing unknown properties to the builder 1`] = `"converters[jsonKey] is not a function"`;
